### ISSUE TITLE
RFC should Magit by default revert buffers visiting tracked files?

### ISSUE THREAD

**Issue by tarsius** [magit/magit#1809]
Okay this is up for discussion again. It seems certain feature areas come up once in a while, and when they do then that happens in clusters of loosely related issues. So should it `magit-auto-revert-mode` be enabled by default?

**Pro**

* It has been enabled for a long time, turning it off would be surprising.
* For users who switch branches, reset, rebase etc. a lot, not having it enabled would be a big step backward.

**Cons**

* When a user *relies* on the buffer not being reverted (as in the buffer not being reverted being part of his workflow) then this can lead to dataloss
* When many buffers are visiting files in the repository, then this can be slow.

**Neither**

I.e. force users to make a choice upfront (or a little later).

* Pro: Since (a) neither having this mode on *or* off is right for everyone, and (b) not every users is aware that they have the option, the *only* way to give everyone what is right for them, is to make them aware that they have a choice.
* Cons: Some users will complain about having to make a choice.

----

Until *very recently* I was of the opinion that the mode should remain **on**, because it has been on for a very long time and because many Magit features are designed with the assumption that it is.

Until *recently* I was of the opinion, that while it should stay on users should **be made aware** of the implications.

Now that the recently opened issues have sunken in I don't really know anymore.

Ps: For the `1.4` series I will probably leave it as is, but now is the time to decide what to do for `2.1.0`.

----

**"Changelog"**

I.e. the issues that in sum lead me to create this one. For context.

* #1783 `magit-auto-revert-mode` should be off by default!!
* #1803 Upgrade warning message is too intrusive; magit-last-seen-setup-instructions
* #1807 Is it practical to limit magit-revert-buffers to known affected buffers?
* #1808 (only marginally related) Magit not refreshing information properly

**Comment by tarsius** [comment]
I currently lean toward "off by default and don't ask". It's the safe choice. It *will* annoy a lot of users, but the only differences is that *different* users are annoyed than if I did something else.

Being reminded that having this mode on can also be a performance issue, is what tipped me over. For performance issues in general I have come to the conclusion that it is best to play it safe; everything that can be slow should be off by default. If a user misses a feature, then she is likely to look/ask if it exists, if Magit is slow, then an unfortunate number of users will just complain without first checking whether something can be done about it.

**Comment by johnmastro** [comment]
Personally I think it makes the most sense to leave it on by default.

The possibility of data loss obviously needs to be taken seriously, but it doesn't seem to have affected many users (which seems to suggest that most existing users understand and want the current behavior). But maybe there's a way to keep the current behavior without risking anyone's data being lost.

If I understood correctly the conversation about data loss, the problem is that the modifications removed by a revert do not end up in the undo list. Maybe it would be possible to ensure that they do, so that someone could always `undo` to their previous state?

I don't see the speed issue as a serious argument for changing the default - people who run into slowness, and are willing to give up the convenience, can disable it in their init file.

**Comment by johnmastro** [comment]
But, to be clear, I won't complain either way, and I hope others won't either :)

**Comment by Silex** [comment]
I'm curious about what workflows requires these buffers not being reverted. IMHO I think only very few people rely on this weird behavior, and the others find logical that auto-revert is on.

If they want that behavior as a mean to have "old version of some file available", this is better done by pressing RET on a file in a commit instead of C-RET.

**Comment by garyo** [comment]
On Thu, Apr 16, 2015 at 3:13 PM, Philippe Vaucher <notifications@github.com>
wrote:

> I'm curious about what workflows requires these buffers not being reverted.


Yes, this seems to be a very questionable workflow in emacs.

I generally am very happy with it on, but since I still find performance
quite slow (especially when I'm on Windows where every git invocation is
slow, for instance) I'm considering turning it off.  Ideally a "surgical"
solution that knew which files could be modified and only reverted buffers
visiting those files would be ideal, but I can see how that's nontrivial.

-- 
Gary

**Comment by tarsius** [comment]
> @johnmastro Maybe it would be possible to ensure that [reverts end up on undo list], so that someone could always undo to their previous state?

I believe this is already the case with the latest Emacs release. But I think the bigger issue is that users who don't expect it to happen, won't notice that it does happen.

> @Silex I'm curious about what workflows requires these buffers not being reverted.

See https://github.com/magit/magit/issues/1783#issuecomment-88876572. It seems that "buffers don't change if I don't tell them too" can be used as a kind of ("cheaper") stash.

> I think only very few people rely on this weird behavior, and the others find logical that auto-revert is on.

I wouldn't call it *weird* behavior. If `git stash` didn't exist, then I think this would be quite reasonable (though dangerous (regardless of whether Magit interferes)) way to "stash" changes. Only Git provides stashes as a substitute and only Magit makes those a pleasure to use. Without those tools you would have to create a copy (another file), or rely on the buffer contents (at least for simple things that don't span across multiple files).

But I fully agree that it is much better to use the tools that Magit/Git provide, obviously.

> I still find performance quite slow (especially when I'm on Windows where every git invocation is slow, for instance) I'm considering turning it off.

I am not sure if this is still true (I don't *know*, I don't use Windows). One of the problems was that on Windows Git comes with a bunch of wrapper scripts which *are* slow - but Magit now uses the absolute path to the actual executable. That should help - maybe not enough.

> @garyo Ideally a "surgical" solution that knew which files could be modified and only reverted buffers visiting those files would be ideal,

It doesn't do that to the extend you are suggesting in #1807, but Magit already only refreshes buffers that might need to be refreshed, i.e. buffers that are inside the repository and visit tracked files that actually changed on disk. Maybe checking these things take to long, this could be benchmarked by splitting them out into separate functions, just to see which parts take long. Maybe `file-readable-p` and `verify-visited-file-modtime` do take a lot of time on Windows?

**Comment by Silex** [comment]
Well, I still find it disconnected reguarding the classic git workflow... I mean, in the console if you checkout/commit/revert the file contents do change. If you modified the file then git refuses to do the operations until you stash or commit your changes.

In the comment you linked he said that he lost some code because of magit-auto-revert, but I have trouble imagining a scenario where that happens. Either he had modifications and then git would have refused, or he didn't and in that case auto-revert does not hurt.

In the vim world, when some process modifies the file you are looking at it asks you wether you want to "Reload this file (y), do nothing (n), reload all modified files (a)". Maybe it's something worth inspiring from... e.g `(setq magit-autorevert-behavior 'always) and with possible value `'ask` or `'never`.

That being said, enabled or not by default, as long as it is configurable probably that any of them is fine :) It's really personal preference, I suspect that much more people expect it on by default but maybe not.

**Comment by npostavs** [comment]
> In the vim world, when some process modifies the file you are looking at it asks you wether you want to

If that's **only** the file you're looking at I guess it's okay. I really like Emacs' don't-change-buffer-until-I-do-something policy much more, and am constantly annoyed by non-Emacs editors that popup *n* dialogs when *n* files are modified.

Not sure if this matters, but `vc-retrieve-tag` (vc.el's change branch equivalent) has `(let ((update (yes-or-no-p "Update any affected buffers? "))`

> I don't see the speed issue as a serious argument for changing the default - people who run into slowness, and are willing to give up the convenience, can disable it in their init file.

Hmm, what if we turn it around: people who are bothered by the inconvenience, and are willing to give up the speed, can enable it in their init file. ;)

---

Would it be too crazy to base the default value on `global-auto-revert-mode`? Or does having `global-auto-revert-mode` on make magit's auto revert redundant?

**Comment by phil-s** [comment]
I'm fully in favour of forcing users to make a choice upfront (or a little later), and consequently ensuring that they're aware of the feature, and how to change it.

It seems that no matter which approach you take, someone will be unhappy about it; but having people unhappy about being made to deal with upgrade/installation instructions (once!) seems *much*  better than having people unhappy because things didn't work they way they expected (or they lost data), with possibly an extended period of not know what the issue is.

**Comment by tarsius** [comment]
Moving from a mode to something with three values `ask`, `always`, and `never` seems like a good idea. I also agree that it shouldn't for each file individually. Something like "Revert n buffers visiting files in this repository? [y]es [n]o [c]: set default".

> Would it be too crazy to base the default value on `global-auto-revert-mode`? Or does having `global-auto-revert-mode` on make magit's auto revert redundant?

No, but in git repositories the magit variant does make the *generic variant* redundant. The latter sits around for five seconds before it does anything and then it reverts asynchronously, while the former kicks in instantly and doesn't stop until it is done. I have experimented with using timers with the magit variant too, but found it to be very annoying (it meant that emacs was waiting to wait to see if it would not get in my way, and me waiting for emacs to do what I asked it to do). On the other hand for people having performance issues, using the generic mode instead of the magit variant might be an option.

**Comment by tarsius** [comment]
@npostavs Regarding `global-auto-revert-mode`. That's still useful, even in git repositories since files might change for any reason. The magit variant only kicks in after invoking a magit command.

> @Silex: Either he had modifications and then git would have refused, or he didn't and in that case auto-revert does not hurt.

I think I came to the same conclusion when this last came up. Unfortunately when it came up again right before I was making a release there was some "you make me lose data" finger pointing, and I did not remember this.

However this assumes that the changes to the file were made using Emacs (either "just emacs" or using magit). When changes are made through other means then the Emacs buffer might indeed hold information no longer in the file.

So we are back where we started, I suppose.

**Comment by tarsius** [comment]
Maybe default to "do it but warn":

```
N files reverted ([C-x u] to undo, [C-c C-c] to customize revert behavior)
```

And the possible behaviors would be:

* revert but warn (and use a transient keymap to allow revert and customize)
* revert but warn (but without instructions and transient key bindings)
* revert silently
* don't revert
* ask (for all affected files at once)

The 99% would still get the expected behavior out of the box. Those annoyed by the message can turn it off. And the 1% cannot accuse me of killing puppies.

Of course it is work to implement this properly. Sight.

**Comment by Silex** [comment]
A common term for this in other editors is `reload`, like `reload from disk?`.

Personally I would not sweat this issue too much :) Just document how to turn magit-revert-mode on or off by default and pick anything for its initial state, people will update their config if it matters to them. There are more important issues to tackle IMHO.

**Comment by sls** [comment]
If there was a way to tell me when I visit a modified buffer that the underlying file has changed, and offer me the chance to accept the filesystem version or not, that would be nice. I'd love to be able to save my changes to another file or compare them with the filesystem version before deciding. 

It's true that it seems odd for a git workflow. Imagine this, though. You are making a change, and something you expected to see isn't there. You look in git and see it was removed yesterday, so you ping the person who did and ask them what's it about. After a brief discussion, they put it back until a better solution can be found. You pull. 

But oh! You weren't thinking and other files you were editing have also changed. You lost sight of this as you were thinking only of getting this fix.

The exact details about who gets notified and how often I leave to others, but I will point out that I got the message when upgrading 29 packages, and it's only by coincidence that I happened to see it on the screen amongst the many warnings generated by the normal run of elisp packages being added. So if it's important that the user see this message, another mechanism may be desirable.

**Comment by tarsius** [comment]
> You pull. But oh!

Either you have saved the file in which case `git pull` will say "I can't do that dave" or you have not saved it in which case magit silently forgoes reverting the buffer.

> So if it's important that the user see this message, another mechanism may be desirable.

I have used the mechanism provided by Emacs to do such things. It has been pointed out that doing that is to intrusive - and now it seems it's not intrusive enough to actually always be noticed...

----

I think I'll go with what I have said here https://github.com/magit/magit/issues/1809#issuecomment-96052241.

**Comment by johnmastro** [comment]
Sounds like a good compromise :+1:

**Comment by tarsius** [comment]
> If there was a way to tell me when I visit a modified buffer that the underlying file has changed,

An "M" will appear somewhere on the left side of the mode-line.

> and offer me the chance to accept the filesystem version or not, that would be nice.

`M-x revert-buffer` vs. do nothing

**Comment by tarsius** [comment]
Yes, it should. I have implemented the features I suggested above in #1846.
